src/number/hash_normalize.cr

```diff
@@ -0,0 +1,110 @@
+  # so that hash(x) == hash(y) whenever x and y are numerically equal,
+  # even if x and y have different types.
+  # A quick summary of the hashing strategy:
+  # (1) First define the 'reduction of x modulo P' for any rational
+  # number x; this is a standard extension of the usual notion of
+  # reduction modulo P for integers.  If x == p/q (written in lowest
+  # terms), the reduction is interpreted as the reduction of p times
+  # the inverse of the reduction of q, all modulo P; if q is exactly
+  # divisible by P then define the reduction to be infinity.  So we've
+  # got a well-defined map
+  #   reduce : { rational numbers } -> { 0, 1, 2, ..., P-1, infinity }.
+  # (2) Now for a rational number x, define hash(x) by:
+  #   reduce(x)   if x >= 0
+  #   -reduce(-x) if x < 0
+  # If the result of the reduction is infinity (this is impossible for
+  # integers, floats and Decimals) then use the predefined hash value
+  # HASH_INF for x >= 0, or -HASH_INF for x < 0, instead.
+  # HASH_INF, -HASH_INF and HASH_NAN are also used for the
+  # hashes of float and Decimal infinities and nans.
+  # A selling point for the above strategy is that it makes it possible
+  # to compute hashes of decimal and binary floating-point numbers
+  # efficiently, even if the exponent of the binary or decimal number
+  # is large.  The key point is that
+  #   reduce(x * y) == reduce(x) * reduce(y) (modulo HASH_MODULUS)
+  # provided that {reduce(x), reduce(y)} != {0, infinity}.  The reduction of a
+  # binary or decimal float is never infinity, since the denominator is a power
+  # of 2 (for binary) or a divisor of a power of 10 (for decimal).  So we have,
+  # for nonnegative x,
+  #   reduce(x * 2**e) == reduce(x) * reduce(2**e) % HASH_MODULUS
+  #   reduce(x * 10**e) == reduce(x) * reduce(10**e) % HASH_MODULUS
+  # and reduce(10**e) can be computed efficiently by the usual modular
+  # exponentiation algorithm.  For reduce(2**e) it's even better: since
+  # P is of the form 2**n-1, reduce(2**e) is 2**(e mod n), and multiplication
+  # by 2**(e mod n) modulo 2**n-1 just amounts to a rotation of bits.
+
+  # XXX: private methods used instead of constants cause current crystal
+  # compiler inlines methods better. Looks like it treats constants as
+  # real variables, and check for their initialization on every access.
+  {% if LibC::ULong == UInt64 %}
+    private HASH_BITS = 61
+    private def hash_bits
+      61
+    end
+    private def int_to_hashnorm(v)
+      v.to_i64
+    end
+  {% else %}
+    private HASH_BITS = 31
+    private def hash_bits
+      31
+    end
+    private def int_to_hashnorm(v)
+      v.to_i32
+    end
+  {% end %}
+
+  private def hash_modulus
+    int_to_hashnorm {{ (1_u64 << HASH_BITS) - 1 }}
+  end
+
+  private def hash_nan
+    int_to_hashnorm 0
+  end
+
+  private def hash_infinity
+    int_to_hashnorm 314159
+  end
+
+  # This function is for reference implementation.
```

Comment:
I improved comment.
I mentioned, that currently hash for BigFloat is not precise for numbers with big fractional part (ie it doesn't distinguish '1.0000000000000001' and '1.0000000000000002').
It happens, cause `BigFloat#hash_normalize` falls back to `to_f.hash_normalize` for such numbers.
It is possible to improve, using this `float_normalize_reference` and overloading `Math.frexp`. (or introducing `BigFloat#frexp` and using it).

Should it be done? (ie, is hashing of such numbers is common task)?
If yes, should it be done in this PR?